src/irmin-pack/unix/sparse_file.ml

```diff
@@ -0,0 +1,335 @@
+
+  let close t =
+    Unix.close t.fd;
+    (* following tries to make the array unreachable, so GC'able; however, no guarantee
+       that arr actually is unreachable *)
+    t.arr <- Bigarray.(Array1.create Int c_layout 0);
+    ()
+end
+
+module Int64_mmap : sig
+  type t = private {
+    fn : string;
+    fd : Unix.file_descr;
+    mutable arr : int64_bigarray;
+  }
+
+  val open_ro : fn:string -> sz:int -> t
+  (** NOTE [open_ ~fn ~sz] can use [sz=-1] to open with size based on the size
+      of the underlying file *)
+
+  val close : t -> unit
+end = struct
+  type t = { fn : string; fd : Unix.file_descr; mutable arr : int64_bigarray }
+
+  (* NOTE sz=-1 is recognized by [map_file] as "derive from size of file"; if we want a
+     different size (eg because we want the file to grow) we can provide it explicitly *)
+  let open_ro ~fn ~sz =
+    let shared = false in
+    assert (Sys.file_exists fn);
+    let fd = Unix.(openfile fn [ O_RDONLY ] 0o660) in
+    let arr =
+      let open Bigarray in
+      Unix.map_file fd Int64 c_layout shared [| sz |] |> array1_of_genarray
+    in
+    { fn; fd; arr }
+
+  let close t =
+    Unix.close t.fd;
+    (* following tries to make the array unreachable, so GC'able; however, no guarantee
+       that arr actually is unreachable *)
+    t.arr <- Bigarray.(Array1.create Int64 c_layout 0);
+    ()
+end
+
+module Make (Io : Io.S) = struct
+  module Io = Io
+  module Errs = Io_errors.Make (Io)
+  module Ao = Append_only_file.Make (Io) (Errs)
+
+  module Mapping_file = struct
+    (** The mapping file is created from a decreasing list of
+        [(virtual_offset, 0, length)]. We first need to reverse it such that
+        virtual offsets are in increasing order. *)
+    let rev_inplace (src : int_bigarray) : unit =
+      let src_sz = BigArr1.dim src in
+      let _ =
+        assert (src_sz >= 3);
+        assert (src_sz mod 3 = 0)
+      in
+      let rec rev i j =
+        if i < j then (
+          let ioff, ilen = (src.{i}, src.{i + 2}) in
+          let joff, jlen = (src.{j}, src.{j + 2}) in
+          src.{i} <- joff;
+          src.{i + 2} <- jlen;
+          src.{j} <- ioff;
+          src.{j + 2} <- ilen;
+          rev (i + 3) (j - 3))
+      in
+      rev 0 (src_sz - 3)
+
+    (** We then replace the [0] component of the triplets with the accumulated
+        length. This yields triplets
+        [(virtual_offset, physical_offset, length)], which will allow us to map
+        virtual offsets to their physical location in the prefix file. *)
+    let set_prefix_offsets src =
+      let src_sz = BigArr1.dim src in
+      let rec go i poff =
+        if i < src_sz then (
+          src.{i + 1} <- poff;
+          let len = src.{i + 2} in
+          go (i + 3) (poff + len))
+      in
+      go 0 0
+
+    type t = { arr : int64_bigarray; root : string; generation : int }
+
+    let open_map ~root ~generation =
+      let path = Irmin_pack.Layout.V4.mapping ~generation ~root in
```

Comment:
Are you planning to update these hard-coded paths in this PR or a future one?

Reply:
Oh thanks I missed it when shuffling the code around!